tidy3d/material_library/parametric_materials.py

```diff
@@ -0,0 +1,266 @@
+        """
+
+        def evaluate(omega: List[float], coeffslist: List[List[float]]) -> List[float]:
+            """Evaluate the Pade approximant given by ``coeffs`` to ``omega``."""
+            res = np.zeros(len(omega), dtype=complex)
+            for coeffs in coeffslist:
+                res += (coeffs[0] + coeffs[1] * 1j * omega) / (
+                    1 + coeffs[2] * 1j * omega + coeffs[3] * (1j * omega) ** 2
+                )
+            return res
+
+        def fit_single(
+            omega: List[float], sigma: List[complex], inds: Tuple[int, int]
+        ) -> List[float]:
+            """Fit a single Pade approximant of degree (1, 2) to ``sigma``
+            as a real function of i ``omega``. The method is described in
+
+            Adam Mock, "Pade approximant spectral fit for FDTD
+            simulation of graphene in the near infrared,"
+            Opt. Mater. Express 2, 6, pp. 771-781 (2012).
+            """
+            gamma = [np.real(sigma[i]) for i in inds]
+            eta = [np.imag(sigma[i]) for i in inds]
+            omegas = [omega[i] for i in inds]
+            matrix = np.array(
+                [
+                    [1, 0, omegas[0] * eta[0], omegas[0] ** 2 * gamma[0]],
+                    [0, omegas[0], -omegas[0] * gamma[0], omegas[0] ** 2 * eta[0]],
+                    [1, 0, omegas[1] * eta[1], omegas[1] ** 2 * gamma[1]],
+                    [0, omegas[1], -omegas[1] * gamma[1], omegas[1] ** 2 * eta[1]],
+                ]
+            )
+            return np.linalg.pinv(matrix) @ np.array([gamma[0], eta[0], gamma[1], eta[1]])
+
+        def optimize(
+            omega: List[float],
+            sigma: List[complex],
+            indslist: List[Tuple[int, int]],
+            coeffslist: List[List[float]],
+        ) -> List[float]:
+            """Optimize the coefficients in ``coeffslist`` by sampling ``omega`` and ``sigma``
+            at the indices in ``indslist``."""
+            for _ in range(num_iters):
+                res = sigma - evaluate(omega, coeffslist)
+                for j, coeffs in enumerate(coeffslist):
+                    curr_res = res + evaluate(omega, [coeffs])
+                    coeffslist[j] = fit_single(omega, curr_res, indslist[j])
+            return coeffslist
+
+        def get_pole_residue(coeffslist: List[List[float]]) -> PoleResidue:
+            """Convert a list of Pade coefficients into a :class:`.PoleResidue` model."""
+            poles = []
+            for coeffs in coeffslist:
+                disc = coeffs[2] ** 2 - 4 * coeffs[3]
+                if disc > 0:
+                    root1 = (coeffs[2] + np.sqrt(disc)) / (2 * coeffs[3])
+                    root2 = (coeffs[2] - np.sqrt(disc)) / (2 * coeffs[3])
+                else:
+                    root1 = (coeffs[2] + 1j * np.sqrt(-disc)) / (2 * coeffs[3])
+                    root2 = (coeffs[2] - 1j * np.sqrt(-disc)) / (2 * coeffs[3])
```

Comment:
is root2 just `np.conj(root1)` in both cases?

Reply:
no, in the second case the roots are real. However, one has very high frequency and could probably be dropped to a good approximation

Reply:
no, in the first case the roots are real. in the second case, they are complex conjugates

Reply:
ok I see. I just remembered something you can do to acommplish the same thing without needing to handle negative `disc`. If you make `disc` complex, bumpy will handle it as you have here 
```py
>>> disc = -55
>>> np.sqrt(complex(disc)) == 1j * np.sqrt(abs(disc))
True
```
so you could replace these 6 lines with
```py
root1 = (coeffs[2] + np.sqrt(complex(disc))) / (2 * coeffs[3])
root2 = (coeffs[2] - np.sqrt(complex(disc))) / (2 * coeffs[3])
```